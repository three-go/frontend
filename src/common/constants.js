const iconNames = {
  heart: "heart",
  brokenHeart: "heart-broken-outline",
  leftArrow: "arrow-left-thick",
  flag: "flag",
};

const iconSizes = {
  headerChance: 28,
  headerLeftArrow: 40,
  headerFlag: 32,
};

const colors = {
  black: "#000000",
  dark: "#212529",
  blueGray: "#2F455C",
  ligthGray: "#5c6977",
  gray: "#96A1A8",
  ivory: "#FCF8F6",
  white: "#ffffff",
  red: "#f45a5a",
  darkRed: "#c92a2a",
  green: "#21D0B2",
  tealGreen: "#00BBD1",
  turquoise: "#00E0EA",
  lightBlue: "#5fceec",
};

const gameStatus = {
  ready: "ready",
  start: "start",
  input: "input",
  end: "end",
};

const navigations = {
  main: "Main",
};

const map = {
  width: 300,
  height: 450,
};

const facePosition = {
  top: 60,
  bottom: 390,
};

const faceYawAngleRange = {
  front: {
    low: -10,
    high: 10,
  },
  left: -36,
  right: 36,
};

export {
  iconNames,
  iconSizes,
  navigations,
  map,
  colors,
  facePosition,
  faceYawAngleRange,
};
