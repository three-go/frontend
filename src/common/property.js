import uuid from "react-native-uuid";

const gameNames = {
  game1: "GAME 1",
  game2: "GAME 2",
};

const gameColors = {
  game1: "#00BBD1",
  game2: "#00E0EA",
};

const game1Guide = [
  {
    title: "게임 설명",
    titleContent:
      "• 시작지점부터 종료지점까지 장애물을 피해 미로를 탈출하는 게임입니다.\n• 총 3개의 stage로 구성되어 있습니다.",
    children: [
      {
        id: uuid.v4(),
        title: "Map",
        content:
          "• 일정시간동안 맵을 기억하세요.\n• 탈출하는 최단거리를 기억하세요",
        image: require("../../public/assets/images/game1/map.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Face",
        content: "• 얼굴로 캐릭터가 움직이는 방향을 기록하세요.",
        image: require("../../public/assets/images/game1/camera.png"),
        icon: "face-recognition",
      },
      {
        id: uuid.v4(),
        title: "Play",
        content: "• 기록된 방향에 따라 캐릭터가 움직입니다.",
        image: require("../../public/assets/images/game1/play.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Next Stage",
        content: "• 캐릭터가 종료지점에 도착하면 다음 스테이지로 이동합니다.",
        image: require("../../public/assets/images/game1/next.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Fail",
        content:
          "• 캐릭터가 종료지점에 도착하지 않으면 목숨 하나가 소진된다.\n• 다시시도 또는 메인화면으로 돌아갈수 있다.",
        image: require("../../public/assets/images/game1/fail.jpg"),
      },
      {
        id: uuid.v4(),
        title: "End",
        content: "• 목숨을 모두 소진하면 게임이 종료된다.",
        image: require("../../public/assets/images/game1/end.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Success",
        content: "• 마지막 stage까지 성공하면 내 점수를 입력할 수 있습니다.",
        image: require("../../public/assets/images/game1/success.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Score",
        content: "• 지금까지 플래이한 점수를 볼 수 있습니다.",
        image: require("../../public/assets/images/game1/score.jpg"),
      },
    ],
  },
  {
    title: "플레이 방법",
    titleContent: "• 게임 진행 방법에 대하여 알려드리겠습니다",
    children: [
      {
        id: uuid.v4(),
        title: "Face",
        content:
          "• 위쪽: 얼굴의 눈 위치를 카메라 화면에 상단에 위치시키세요\n• 아래쪽: 얼굴의 눈 위치를 카메라 화면에 하단에 위치시키세요\n• 왼쪽: 얼굴의 고개를 왼쪽으로 돌리세요.\n• 오른쪽: 얼굴의 고개를 오른쪽으로 돌리세요.",
        image: require("../../public/assets/images/game1/camera.png"),
        icon: "face-recognition",
      },
      {
        id: uuid.v4(),
        title: "Play",
        content: "• 기록된 방향에따라 캐릭터가 자동으로 움직입니다.",
        image: require("../../public/assets/images/game1/play.jpg"),
      },
    ],
  },
];

const gameDescription = {
  game1: game1Guide,
  game2: [],
};

const gameKey = ["game1", "game2"];

const game = {
  keys: gameKey,
  names: gameNames,
  colors: gameColors,
  description: gameDescription,
};

const colors = {
  ivory: "#FCF8F6",
};

export { game, colors };
