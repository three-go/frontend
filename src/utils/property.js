const gameNames = {
  game1: "GAME 1",
  game2: "GAME 2",
};

const gameColors = {
  game1: "#00BBD1",
  game2: "#00E0EA",
};

const gameDescription = [["게임 설명 1111"], ["게임 설명 2222"]];

const gameKey = ["game1", "game2"];

const game = {
  keys: gameKey,
  names: gameNames,
  colors: gameColors,
  description: gameDescription,
};

export { game };
