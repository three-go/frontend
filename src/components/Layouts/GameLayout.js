import React from "react";

import { useNavigation } from "@react-navigation/native";
import { View, StyleSheet } from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";

import { GameHeader } from "..";
import { navigations, colors } from "../../common";

const GameLayout = ({
  children,
  status,
  setStatus,
  startTimer,
  setStartTimer,
  inputTimer,
  setInputTimer,
  cameraPermissionStatus,
  score,
  currentGameKey,
  progressRate,
  setProgressRate,
  chance,
}) => {
  const navigation = useNavigation();

  const handleGoBack = () => {
    navigation.navigate(navigations.main);
  };

  return (
    <SafeAreaView style={styles.container} edges={["top"]}>
      <View style={styles.header}>
        <GameHeader
          status={status}
          setStatus={setStatus}
          onPressBack={handleGoBack}
          startTimer={startTimer}
          setStartTimer={setStartTimer}
          inputTimer={inputTimer}
          setInputTimer={setInputTimer}
          cameraPermissionStatus={cameraPermissionStatus}
          score={score}
          currentGameKey={currentGameKey}
          progressRate={progressRate}
          setProgressRate={setProgressRate}
          chance={chance}
        />
      </View>
      <View style={styles.gameBoard}>{children}</View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.dark,
  },
  header: {
    height: "10%",
  },
  gameBoard: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    width: "100%",
    height: "100%",
    backgroundColor: colors.ivory,
  },
});

export default GameLayout;
