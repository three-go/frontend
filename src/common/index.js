import {
  iconNames,
  iconSizes,
  navigations,
  map,
  colors,
  facePosition,
  faceYawAngleRange,
} from "./constants";
import { game } from "./property";

export {
  iconNames,
  iconSizes,
  navigations,
  map,
  colors,
  facePosition,
  faceYawAngleRange,
  game,
};
