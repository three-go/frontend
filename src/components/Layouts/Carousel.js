import React from "react";

import { FlatList, StyleSheet, View } from "react-native";
import PropTypes from "prop-types";

import { game } from "../../common/property";

const Carousel = ({ content, renderItem, size }) => {
  return (
    <View style={styles.container}>
      <FlatList
        automaticallyAdjustContentInsets={false}
        contentContainerStyle={styles.listWrapper(size)}
        data={content}
        decelerationRate="fast"
        horizontal
        keyExtractor={(_, index) => game.keys[index]}
        pagingEnabled
        renderItem={renderItem}
        snapToInterval={size.PAGE_WIDTH + size.GAP}
        snapToAlignment="start"
        showsHorizontalScrollIndicator={false}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    height: "60%",
    justifyContent: "center",
    alignItems: "center",
  },
  listWrapper: (prop) => {
    return {
      paddingHorizontal: prop.OFFSET + prop.GAP / 2,
    };
  },
});

export default Carousel;

Carousel.propTypes = {
  content: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.shape({
        children: PropTypes.array,
        title: PropTypes.string,
        titleContent: PropTypes.string,
      }),
      PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.string,
          name: PropTypes.string,
          score: PropTypes.number,
        })
      ),
    ])
  ).isRequired,
  renderItem: PropTypes.func.isRequired,
  size: PropTypes.shape({
    GAP: PropTypes.number,
    OFFSET: PropTypes.number,
    PAGE_WIDTH: PropTypes.number,
    SCREEN_WIDTH: PropTypes.number,
  }).isRequired,
};
