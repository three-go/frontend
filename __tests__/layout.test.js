import React from "react";

import { render, fireEvent } from "@testing-library/react-native";
import { Text, View } from "react-native";

import "@testing-library/jest-native";
import "react-native-gesture-handler/jestSetup";
import { camera } from "../src/common/constants";
import { game } from "../src/common/property";
import Carousel from "../src/components/Layouts/Carousel";
import GameLayout from "../src/components/Layouts/GameLayout";

const mockNavigate = jest.fn();
const mockDisPatch = jest.fn();

jest.mock("react-native-sound", () => {
  const Sound = () => {
    const obj = {};
    obj.sound = "music";
    return obj;
  };

  return Sound;
});

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockNavigate,
      dispatch: mockDisPatch,
    }),
  };
});

describe("Layout Test", () => {
  it("Carousel Test", () => {
    const props = {
      content: [
        { title: "Title1", description: "description1" },
        { title: "Title2", description: "description2" },
      ],
      renderItem: ({ item }) => {
        return (
          <View>
            <Text>{item.title}</Text>
            <Text>{item.description}</Text>
          </View>
        );
      },
      size: {
        SCREEN_WIDTH: 300,
        GAP: 15,
        OFFSET: 15,
        PAGE_WIDTH: 300 - (15 + 15) * 2,
      },
    };

    const carousel = render(<Carousel {...props} />);
    const carouselJson = carousel.toJSON();

    expect(carousel.getByText("Title1")).toBeTruthy();
    expect(carousel.getByText("description1")).toBeTruthy();
    expect(carousel.getByText("Title2")).toBeTruthy();
    expect(carousel.getByText("description2")).toBeTruthy();
    expect(carouselJson).toMatchSnapshot();
    expect(carouselJson).toBeTruthy();
  });

  it("GameLayout Test", () => {
    const props = {
      status: "mapOpen",
      setStatus: jest.fn(),
      chance: 3,
      score: 0,
      cameraPermissionStatus: camera.permissionNotAuthorized,
      currentGameKey: game.keys[0],
    };

    const gameLayout = render(
      <GameLayout {...props}>
        <Text>children component</Text>
      </GameLayout>
    );
    const gameLayoutJson = gameLayout.toJSON();
    const backButton = gameLayout.getByTestId("backButton");
    fireEvent.press(backButton);

    expect(gameLayout.getByText("children component")).toBeTruthy();
    expect(gameLayoutJson).toMatchSnapshot();
    expect(gameLayoutJson).toBeTruthy();
    expect(mockNavigate).toHaveBeenCalledTimes(1);
  });
});
