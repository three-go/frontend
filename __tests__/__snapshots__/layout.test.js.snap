// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Test Carousel Test 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "height": "60%",
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentContainerStyle={
      Object {
        "paddingHorizontal": 22.5,
      }
    }
    data={
      Array [
        Object {
          "description": "description1",
          "title": "Title1",
        },
        Object {
          "description": "description2",
          "title": "Title2",
        },
      ]
    }
    decelerationRate="fast"
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    pagingEnabled={true}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    snapToAlignment="start"
    snapToInterval={255}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View>
          <Text>
            Title1
          </Text>
          <Text>
            description1
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View>
          <Text>
            Title2
          </Text>
          <Text>
            description2
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Layout Test GameLayout Test 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "top",
    ]
  }
  style={
    Object {
      "backgroundColor": "#212529",
      "flex": 1,
    }
  }
  testID="layout"
>
  <View
    style={
      Object {
        "height": "10%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#212529",
          "color": "#FCF8F6",
          "flexDirection": "row",
          "height": "100%",
          "justifyContent": "space-evenly",
          "paddingLeft": 8,
          "paddingRight": 8,
          "width": "100%",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "height": 40,
            "width": 65,
          }
        }
        testID="backButton"
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#96A1A8",
                "fontSize": 40,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰁎
        </Text>
      </View>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "marginBottom": 8,
            }
          }
          testID="timer"
        >
          <Text
            style={
              Object {
                "color": "#FCF8F6",
                "fontSize": 12,
              }
            }
          >
            맵이 가려지기
            <Text
              style={
                Object {
                  "color": "#FCF8F6",
                  "fontSize": 24,
                }
              }
            >
              5
            </Text>
            초 전
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#f45a5a",
                "fontSize": 28,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰋑
        </Text>
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#f45a5a",
                "fontSize": 28,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰋑
        </Text>
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#f45a5a",
                "fontSize": 28,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰋑
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FCF8F6",
        "flex": 1,
        "height": "100%",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <Text>
      children component
    </Text>
  </View>
</RNCSafeAreaView>
`;
